### PR TITLE
Add option to listen resource and refactor option storage

### PR DESCRIPTION
**Pull request by bmhughes**
# Description

- Add option to listen resource and refactor option storage

## Issues Resolved

- Replaces #470 

## Check List

- [x] A summary of changes made is included in the CHANGELOG under `## Unreleased`
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable.

### CODE REVIEW

**Review by damacus — DISMISSED** [review]

File: test/integration/config_4/example_spec.rb [hunk 1]
```diff
@@ -0,0 +1,104 @@
+describe package('haproxy') do
```
Comment (damacus) [inline test/integration/config_4/example_spec.rb:1]: Do we need example4 spec?

Comment (bmhughes) [reply test/integration/config_4/example_spec.rb:1]: Nothing else has a `listen` setup to test out the options property. What are we thinking for removing it? Just to get the number of tests down?

Comment (damacus) [reply test/integration/config_4/example_spec.rb:1]: Yeah, happy to move it into one of the other suites so it gets tested.

**Review by Stromweld — DISMISSED** [review]

**Comment by kitchen-porter** [comment]
Released as: [12.4.0](https://github.com/sous-chefs/haproxy/releases/tag/12.4.0)

**Pull request merged**